src/core/server/workspaces/workspace_permission_check.ts

```diff
@@ -0,0 +1,273 @@
+export const PERMISSION_TYPE_WRITE = 'write';
+export const PERMISSION_TYPE_MANAGEMENT = 'management';
+export const PERMISSION_TYPE_LIBRARY_READ = 'library_read';
+export const PERMISSION_TYPE_LIBRARY_WRITE = 'library_write';
+
+export const PERMISSION_TYPES: Set<string> = new Set([
+  PERMISSION_TYPE_READ,
+  PERMISSION_TYPE_WRITE,
+  PERMISSION_TYPE_MANAGEMENT,
+  PERMISSION_TYPE_LIBRARY_READ,
+  PERMISSION_TYPE_LIBRARY_WRITE,
+]);
+
+const transferPermissionsToMap = (permissions: Permissions) => {
+  const map: Map<string, Principals> = new Map();
+  if (!!permissions.read) {
+    map.set(PERMISSION_TYPE_READ, permissions.read);
+  }
+  if (!!permissions.write) {
+    map.set(PERMISSION_TYPE_WRITE, permissions.write);
+  }
+  if (!!permissions.management) {
+    map.set(PERMISSION_TYPE_MANAGEMENT, permissions.management);
+  }
+  if (!!permissions.library_read) {
+    map.set(PERMISSION_TYPE_LIBRARY_READ, permissions.library_read);
+  }
+  if (!!permissions.library_write) {
+    map.set(PERMISSION_TYPE_LIBRARY_WRITE, permissions.library_write);
+  }
+  return map;
+};
+
+const addPermissionToPrincipals = (
+  principals?: Principals,
+  users?: string[],
+  groups?: string[]
+) => {
+  if (!principals) {
+    principals = {};
+  }
+  if (!!users && !(principals.users?.length === 1 && principals.users[0] === '*')) {
+    if (!principals.users) {
+      principals.users = [] as string[];
+    }
+    principals.users = principals.users.concat(
+      users.filter((item) => !principals?.users?.includes(item))
+    );
+  }
+  if (!!groups && !(principals.groups?.length === 1 && principals.groups[0] === '*')) {
+    if (!principals.groups) {
+      principals.groups = [] as string[];
+    }
+    principals.groups = principals.groups.concat(
+      groups.filter((item) => !principals?.groups?.includes(item))
+    );
+  }
+  return principals;
+};
+
+const deletePermissionFromPrincipals = (
+  principals?: Principals,
+  users?: string[],
+  groups?: string[]
+) => {
+  if (!principals) {
+    return principals;
+  }
+  if (!!users && !!principals.users) {
+    principals.users = principals.users.filter((item) => !users.includes(item));
+  }
+  if (!!groups && !!principals.groups) {
+    principals.groups = principals.groups.filter((item) => !groups.includes(item));
+  }
+  return principals;
+};
+
+const isParamsValid = (
+  permissonType: string,
+  user?: string | string[],
+  group?: string | string[]
+) => {
+  if ((!user && !group) || !permissonType || !PERMISSION_TYPES.has(permissonType)) {
+    return false;
+  }
+  return true;
+};
+
+export class WorkspacePermissionCheck {
+  public hasPermission(
+    permissonType: string,
+    permissions: Permissions,
+    user?: string,
+    group?: string
+  ) {
+    if (!isParamsValid(permissonType, user, group) || !permissions) {
+      return false;
+    }
+    const permissionMap = transferPermissionsToMap(permissions);
+    const principals = permissionMap.get(permissonType);
```

Comment:
Q: is it necessary to convert it to a `Map` first? we can simply do `const principals = permissions[permissonType]`, isn't it?

Reply:
That is doable, I've changed that.